File: combinatorics/resource/ST/project_6/dijkstra_general.py
import heapq

# Ham Dijkstra tren do thi tong quat
def dijkstra(n, edges, start):
    graph = [[] for _ in range(n)]  # Danh sach ke voi trong so

    # Xay dung do thi
    for u, v, w in edges:
        graph[u].append((v, w))  # Do thi co huong
        # graph[v].append((u, w))  # Neu do thi vo huong

    dist = [float('inf')] * n     # Khoang cach ban dau la vo cung
    dist[start] = 0               # Khoang cach tu start den chinh no = 0

    pq = [(0, start)]             # Hang doi uu tien voi (khoang cach, dinh)

    while pq:
        d, u = heapq.heappop(pq)
        if d > dist[u]: continue  # Neu da co duong ngan hon thi bo qua

        for v, w in graph[u]:
            if dist[v] > dist[u] + w:
                dist[v] = dist[u] + w
                heapq.heappush(pq, (dist[v], v))

    return dist

# Nhap du lieu
n = int(input("Nhap so dinh: "))
m = int(input("Nhap so canh: "))
edges = []
print("Nhap cac canh u v w:")
for _ in range(m):
    u, v, w = map(int, input().split())
    edges.append((u, v, w))

s = int(input("Nhap dinh bat dau: "))
res = dijkstra(n, edges, s)

print("Khoang cach ngan nhat tu dinh", s)
for i, d in enumerate(res):
    print(f"{s} -> {i}: {d}")

File: combinatorics/resource/ST/project_6/dijkstra_general.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
using namespace std;

// Dijkstra tren do thi tong quat
vector<int> dijkstra(int n, const vector<tuple<int, int, int>>& edges, int start) {
    vector<vector<pair<int, int>>> graph(n); // graph[u] = (v, w)

    // Xay dung do thi
    for (auto [u, v, w] : edges) {
        graph[u].emplace_back(v, w);  // Do thi co huong
        // graph[v].emplace_back(u, w); // Neu do thi vo huong
    }

    vector<int> dist(n, INT_MAX);     // Ban dau gan vo cung
    dist[start] = 0;

    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>> pq;
    pq.emplace(0, start);  // (khoang cach, dinh)

    while (!pq.empty()) {
        auto [d, u] = pq.top(); pq.pop();
        if (d > dist[u]) continue;

        for (auto [v, w] : graph[u]) {
            if (dist[v] > dist[u] + w) {
                dist[v] = dist[u] + w;
                pq.emplace(dist[v], v);
            }
        }
    }

    return dist;
}

int main() {
    int n, m;
    cout << "Nhap so dinh va so canh: ";
    cin >> n >> m;

    vector<tuple<int, int, int>> edges;
    cout << "Nhap cac canh u v w:\n";
    for (int i = 0; i < m; ++i) {
        int u, v, w;
        cin >> u >> v >> w;
        edges.emplace_back(u, v, w);
    }

    int s;
    cout << "Nhap dinh bat dau: ";
    cin >> s;

    vector<int> dist = dijkstra(n, edges, s);
    cout << "Khoang cach ngan nhat tu " << s << ":\n";
    for (int i = 0; i < n; ++i) {
        cout << s << " -> " << i << ": " << dist[i] << '\n';
    }

    return 0;
}